### DISCUSSION TITLE
question about QueuePool limit

### DISCUSSION THREAD

**Discussion by 841020** [sqlalchemy/sqlalchemy#6463] [Usage Questions]
**Describe your question**
my project has a [QueuePool limit](https://docs.sqlalchemy.org/en/14/errors.html#connections-and-transactions) issue currently

my question is wht does [“do whatever it is you do with connections when we're done using them”](https://docs.sqlalchemy.org/en/14/glossary.html#term-released) mean?
Does this mean that even if it is just a query, there is no commit (create/update/delete). Close() is required at the end?

**Example - please use the [Minimal, Complete, and Verifiable](http://stackoverflow.com/help/mcve) guidelines if possible**

**Complete stack trace, if applicable**

**Versions**
 - OS:unix
 - Python:3.8.9
 - SQLAlchemy:1.4.7
 - Database:MySQL
 - DBAPI:PyMySQL1.0.2

**Additional context**
![14030841205661](https://user-images.githubusercontent.com/33141341/117752657-bee57e80-b249-11eb-827d-20cec8330176.jpg)
**Useful links**
- [Get Support](https://www.sqlalchemy.org/support.html) on the website
- The [documentation](https://docs.sqlalchemy.org/en/latest/) website
- The [UsageRecipes](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) wiki
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sqlalchemy) tag
- SQLAlchemy [Google group](http://groups.google.com/group/sqlalchemy)
- [Gitter](https://gitter.im/sqlalchemy/community) chat

**Have a nice day!**

**Comment by zzzeek** [comment]
> **Describe your question**
> my project has a [QueuePool limit](https://docs.sqlalchemy.org/en/14/errors.html#connections-and-transactions) issue currently
> 
> my question is wht does [“do whatever it is you do with connections when we're done using them”](https://docs.sqlalchemy.org/en/14/glossary.html#term-released) mean?

the paragraph following tries to describe this:

> The term will sometimes be used in the phrase, “release transactional resources”, to indicate more explicitly that what we are actually “releasing” is any transactional state which as accumulated upon the connection. In most situations, the process of selecting from tables, emitting updates, etc. acquires isolated state upon that connection as well as potential row or table locks. This state is all local to a particular transaction on the connection, and is released when we emit a rollback. An important feature of the connection pool is that when we return a connection to the pool, the connection.rollback() method of the DBAPI is called as well, so that as the connection is set up to be used again, it’s in a “clean” state with no references held to the previous series of operations.

concretely it means:

1. if the pool has space for pooled connections, that is, it's not in overflow, the steps are:
a. call connection.rollback()
b. possibly do some logging, call connection return events
c. put the connection back in the pool

2. if the pool is full and this connection is from "overflow", the steps are:
a. call connection.rollback()
b. possibly do some logging, call connection return events
c. call connection.close()
d. dereference the connection



> Does this mean that even if it is just a query, there is no commit (create/update/delete). Close() is required at the end?

yes, whenever you .connect(), you should .close().    it's best to use a context manager as described at https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html#getting-a-connection .  

While there is a system by which if you just "forget" about the connection and let it be garbage collected, the pool will intercept the connection being garbage collected and do a regular pool checkin.  but this is not a good practice, as it means the connection might be returned to the pool in the GC thread, or not closed soon enough leading to deadlocks, or if you are using asyncio it can't work at all.

**Reply by 841020** [reply]
Thanks you very much
this is all i need to know
I appreciate it
